### PR TITLE
[infra] Add support for dataflow builds to the helper script and build check (#1632).

### PR DESCRIPTION
**Pull request by Dor1s**
PTAL, but do not merge. I'll remove `zstd` change from this PR. Just want to see how Travis handles it as of now.

### CODE REVIEW

File: infra/base-images/base-runner/bad_build_check [hunk 1]
```diff
@@ -89,6 +94,10 @@ function check_engine {
       cat $FUZZER_OUTPUT
       return 1
     fi
+  elif [[ "$FUZZING_ENGINE" == dataflow ]]; then
+    # TODO(https://github.com/google/clusterfuzz/issues/503): add check for
```
Comment (Dor1s) [inline infra/base-images/base-runner/bad_build_check]: replace with https://github.com/google/oss-fuzz/issues/1632

File: infra/base-images/base-runner/bad_build_check [hunk 2]
```diff
@@ -113,6 +122,10 @@ function check_startup_crash {
     if [ $(egrep "target binary (crashed|terminated)" -c $FUZZER_OUTPUT) -eq 0 ]; then
       CHECK_PASSED=1
     fi
+  elif [[ "$FUZZING_ENGINE" = dataflow ]]; then
+    # TODO(https://github.com/google/clusterfuzz/issues/503): add check for
```
Comment (Dor1s) [inline infra/base-images/base-runner/bad_build_check]: and this

File: infra/helper.py [hunk 3]
```diff
@@ -160,6 +160,16 @@ def main():
                                              help='Pull base images.')
 
   args = parser.parse_args()
+  try:
```
Comment (Dor1s) [inline infra/helper.py]: add a comment

File: projects/zstd/project.yaml [hunk 4]
```diff
@@ -4,7 +4,12 @@ auto_ccs:
   - "Yann.collet.73@gmail.com"
   - "terrelln@fb.com"
   - "cyan@fb.com"
+fuzzing_engines:
+  - libfuzzer
+  - afl
+  - dataflow
```
Comment (jonathanmetzman) [inline projects/zstd/project.yaml]: I don't like that someone must specify "dataflow" for fuzzing_engine and sanitizers, maybe we can simply ignore the sanitizer argument when engine dataflow is used instead? Or how about using a "none" sanitizer

File: infra/helper.py [hunk 5]
```diff
@@ -160,6 +160,18 @@ def main():
...
   args = parser.parse_args()
+
+  # We have different default values for `sanitizer` depending on the `engine`.
+  try:
+    if not args.sanitizer:
+      if args.engine == 'dataflow':
+        args.sanitizer = 'dataflow'
+      else:
+        args.sanitizer = 'address'
+  except:
```
Comment (jonathanmetzman) [inline infra/helper.py]: Why is this needed now?

File: infra/helper.py [hunk 6]
```diff
@@ -268,7 +280,11 @@ def _add_sanitizer_args(
         parser,
         choices=('address', 'memory', 'undefined', 'coverage', 'dataflow')):
   """Add common sanitizer args."""
-  parser.add_argument('--sanitizer', default='address', choices=choices)
+  parser.add_argument(
+      '--sanitizer',
+      default=None,
+      choices=choices,
+      help='the default is "address"; "dataflow" for "dataflow" engine')
```
Comment (jonathanmetzman) [inline infra/helper.py]: Not sure we want to document this behavior for "dataflow", maybe remove the help string.

File: infra/base-images/base-runner/bad_build_check [hunk 7]
```diff
@@ -131,15 +144,53 @@ function check_startup_crash {
 function check_asan_build {
   local FUZZER=$1
   local ASAN_CALLS=$2
-  local MSAN_CALLS=$3
```
Comment (jonathanmetzman) [inline infra/base-images/base-runner/bad_build_check]: Nit: It looks like the order of these params were changed to be in alphabetical order, not sure this is worth doing (in fact, I think it is harmful when we adopt this convention in places where it isn't helpful and where obeying simply makes code more difficult to modify).

File: infra/base-images/base-runner/bad_build_check [hunk 8]
```diff
@@ -113,6 +122,10 @@ function check_startup_crash {
     if [ $(egrep "target binary (crashed|terminated)" -c $FUZZER_OUTPUT) -eq 0 ]; then
       CHECK_PASSED=1
     fi
+  elif [[ "$FUZZING_ENGINE" = dataflow ]]; then
+    # TODO(https://github.com/google/oss-fuzz/issues/1632): add check for
+    # binaries compiled with dataflow engine when the interface becomes stable.
+    echo "Skipping check_startup_crash for dataflow engine build."
```
Comment (jonathanmetzman) [inline infra/base-images/base-runner/bad_build_check]: I don't printing this line is helpful, same for previous case, please remove.

File: infra/base-images/base-runner/bad_build_check [hunk 9]
```diff
@@ -27,6 +27,11 @@ THRESHOLD_FOR_NUMBER_OF_EDGES=100
 ASAN_CALLS_THRESHOLD_FOR_ASAN_BUILD=1000
 ASAN_CALLS_THRESHOLD_FOR_NON_ASAN_BUILD=0
 
+# The value below can definitely be higher (like 500-1000), but avoid being too
+# agressive here while still evaluating the DFT-based fuzzing approach.
+DFSAN_CALLS_THRESHOLD_FOR_DFSAN_BUILD=100
```
Comment (jonathanmetzman) [inline infra/base-images/base-runner/bad_build_check]: I think bad_build_check is being too complicated for bash and should be converted to python. To compensate for this, the code is quite repetitive.

File: infra/base-images/base-runner/bad_build_check [hunk 10]
```diff
@@ -233,17 +297,21 @@ function check_mixed_sanitizers {
...
   local MSAN_CALLS=$(objdump -dC $FUZZER | egrep "${CALL_INSN}__msan" -c)
   local UBSAN_CALLS=$(objdump -dC $FUZZER | egrep "${CALL_INSN}__ubsan" -c)
 
   if [[ "$SANITIZER" = address ]]; then
-    check_asan_build $FUZZER $ASAN_CALLS $MSAN_CALLS $UBSAN_CALLS
+    check_asan_build $FUZZER $ASAN_CALLS $DFSAN_CALLS $MSAN_CALLS $UBSAN_CALLS
+    result=$?
+  elif [[ "$SANITIZER" = dataflow ]]; then
+    check_dfsan_build $FUZZER $ASAN_CALLS $DFSAN_CALLS $MSAN_CALLS $UBSAN_CALLS
     result=$?
```
Comment (jonathanmetzman) [inline infra/base-images/base-runner/bad_build_check]: Could you do me favor by hardcoding return nonzero and seeing if check_build fails? I think it doesn't which is a problem.

File: projects/zstd/project.yaml (hunk 4 above)
Comment (Dor1s) [reply projects/zstd/project.yaml]: Yeah, I'm also not quite happy with this, but otherwise I'd have to hack something in the travis script, like make `sanitizer` value optional, then invoke `helper.py` without `--sanitizer` and let it use the default value. Don't like going that path either, as dataflow is actually both a new engine (we link against `DataFlow.o`) and a new sanitizer (we use different compilation flags), so it feels like specifying `dataflow` twice in `project.yaml` is the lesser of the evils.

File: infra/helper.py (hunk 5 above)
Comment (Dor1s) [reply infra/helper.py]: Because some commands do not have `sanitizer` argument and line 166 throws an exception.

File: infra/base-images/base-runner/bad_build_check (hunk 10 above)
Comment (Dor1s) [reply infra/base-images/base-runner/bad_build_check]: Yeah, this new build check is not active until merged and the base images re-built. Right now the check for dataflow build passes become it doesn't do anything.

File: infra/base-images/base-runner/bad_build_check (hunk 9 above)
Comment (Dor1s) [reply infra/base-images/base-runner/bad_build_check]: Agreed, would be nice to have a new implementation, but not in this CL :)

File: infra/helper.py (hunk 5 above)
Comment (jonathanmetzman) [reply infra/helper.py]: Which commands? Why is this a new development?

File: infra/base-images/base-runner/bad_build_check (hunk 8 above)
Comment (Dor1s) [reply infra/base-images/base-runner/bad_build_check]: Hmmm, you're right, that line will be visible in a case of failure, which shouldn't happen unless there is some actual check added. Oh, and I also need to do `CHECK_PASSED=1`. Done.

File: infra/base-images/base-runner/bad_build_check (hunk 10 above)
Comment (jonathanmetzman) [reply infra/base-images/base-runner/bad_build_check]: Err...that's not quite what I mean, I mean that if this script returns nonzero, that nothing happens.
Nevermind since this won't be easy for you to test.

File: infra/base-images/base-runner/bad_build_check (hunk 7 above)
Comment (Dor1s) [reply infra/base-images/base-runner/bad_build_check]: I definitely share your point that sometimes it becomes too annoying, but it feels like we've used to follow that convention in CF / OSS-Fuzz code, so I've just done the same here as well.

File: infra/helper.py (hunk 6 above)
Comment (Dor1s) [reply infra/helper.py]: Can change to `the default is "address", unless using "dataflow" engine`, does that sound better?

File: infra/helper.py (hunk 5 above)
Comment (Dor1s) [reply infra/helper.py]: e.g. `build_image` doesn't require `--sanitizer`:

```
$ python infra/helper.py build_image --no-pull zstd
Traceback (most recent call last):
  File "infra/helper.py", line 852, in <module>
    sys.exit(main())
  File "infra/helper.py", line 166, in main
    if not args.sanitizer:
AttributeError: 'Namespace' object has no attribute 'sanitizer'
```

It wasn't an issue as we previously wouldn't attempt to access `args.sanitizer` unless we knew it's specified (e.g. when `build_fuzzers` command is used). Now I have to access it to assign the default value dynamically. Unfortunately `argparse` doesn't seem to have any safe method like `.get(key)` to check if the value is present or not :|

**Comment by Dor1s** [comment]
> What happens if you modify a project.yaml of a project without dfsan enabled? Does travis skip building it?

I think it should just skip, as `dataflow` isn't specified in the default configurations list and `is_enabled` function in the travis script will return false.

File: infra/base-images/base-runner/bad_build_check (hunk 10 above)
Comment (Dor1s) [reply infra/base-images/base-runner/bad_build_check]: I test this locally by doing:

```bash
python infra/helper.py build_image --no-pull base-runner && ./bootstrap.sh zstd && python infra/helper.py check_build --engine dataflow zstd
```

where `bootstrap.sh` is:

```
#!/bin/bash
if [[ $# -eq 0 ]] ; then
  echo "Usage: $0 <project_name>"
  exit 1
fi

project_name=$1

python infra/helper.py build_image --no-pull $project_name \
&& python infra/helper.py build_fuzzers --engine dataflow --sanitizer dataflow $project_name
```

And have caught some issues, maybe that's what you meant. Fixed those in the latest commit, testing again.

Comment (Dor1s) [reply infra/base-images/base-runner/bad_build_check]: Yeah, now it works.

Comment (Dor1s) [reply infra/base-images/base-runner/bad_build_check]: Resolving this comment branch as well, please reopen if I'm still missing something important.

File: projects/zstd/project.yaml (hunk 4 above)
Comment (Dor1s) [reply projects/zstd/project.yaml]: //cc @oliverchang, in case he's got a better idea on how to indicate a new engine and new sanitizer at once, just discussed this with him

File: infra/helper.py (hunk 5 above)
Comment (jonathanmetzman) [reply infra/helper.py]: What I mean is that this script makes a lot of effort to exit with a nonzero return code if any command it runs exits with nonzero. But the exit value of this script [is ignored](https://github.com/google/oss-fuzz/blob/4d3073292ae43d99b1bd34b58e45021cb83760ae/infra/base-images/base-runner/test_all#L59) in favor of hackily looking for output. Can you leave a TODO to fix this please?

File: projects/zstd/project.yaml (hunk 4 above)
Comment (jonathanmetzman) [reply projects/zstd/project.yaml]: Why not just make a "none" sanitizer and use that in travis?

File: infra/helper.py (hunk 5 above)
Comment (jonathanmetzman) [reply infra/helper.py]: Btw, you can probably use `hasattr` instead of the current approach of catching an exception.

File: infra/base-images/base-runner/bad_build_check (hunk 10 above)
Comment (jonathanmetzman) [reply infra/base-images/base-runner/bad_build_check]: Sorry I meant that (I think) this script tries to plumb nonzero exitcodes, even though those [are ignored.](https://github.com/google/oss-fuzz/blob/4d3073292ae43d99b1bd34b58e45021cb83760ae/infra/base-images/base-runner/test_all#L59), can you please leave a TODO to fix this?

File: projects/zstd/project.yaml (hunk 4 above)
Comment (Dor1s) [reply projects/zstd/project.yaml]: I think `none` sanitizer would be incorrect, as one would expect it not to have any sanitizer flags while in fact it enables several compiler flags: https://github.com/google/oss-fuzz/blob/c60fcd7247da0505e3977b0725d795a96966b3db/infra/base-images/base-builder/Dockerfile#L38

So it's clearly a new sanitizer in our model, it's just not compatible with other fuzzing engines right now, like MSan is not compatible/used with AFL, etc.

File: infra/helper.py (hunk 5 above)
Comment (Dor1s) [reply infra/helper.py]: Oooh, nice! Thanks! Done.

File: projects/zstd/project.yaml (hunk 4 above)
Comment (jonathanmetzman) [reply projects/zstd/project.yaml]: Shouldn't these be coverage flags rather than sanitizer flags? This seems easy to answer for "-fsanitize-coverage" flags, I think its fine too for "-fsanitize=dataflow" since it seems to have more in common with coverage flags than other kinds of sanitizer flags IMO

File: infra/helper.py [hunk 11]
```diff
@@ -162,15 +162,12 @@ def main():
...
-    if not args.sanitizer:
-      if args.engine == 'dataflow':
-        args.sanitizer = 'dataflow'
-      else:
-        args.sanitizer = 'address'
-  except:
-    # Some commands do not have `sanitizer` argument.
-    pass
+  # Some commands do not have `sanitizer` argument, so `hasattr` is necessary.
+  if hasattr(args, 'sanitizer') and not args.sanitizer:
```
Comment (jonathanmetzman) [inline infra/helper.py]: nit: Sorry should've suggested `if not getattr(args, 'sanitizer', None)` as it is shorter.

File: projects/zstd/project.yaml (hunk 4 above)
Comment (oliverchang) [reply projects/zstd/project.yaml]: I think having both engine and sanitizer is fine here, since that's actually what's happening here. Hacking around it most likely won't be very nice.

+1 to putting the coverage flags in the right place, but I'm not sure about -fsanitize=dataflow.

File: infra/helper.py (hunk 11 above)
Comment (jonathanmetzman) [reply infra/helper.py]: If I lose this disagreement over whether a seperate engine exists then please fail loudly if --sanitizer dataflow is specified but engine isn't data flow.

Comment (Dor1s) [reply infra/helper.py]: I think this won't be that useful, as I need to distinguish whether the argument is not used at all (the first condition) or is not specified (the second condition) and needs a default value.

File: infra/base-images/base-runner/bad_build_check (hunk 10 above)
Comment (Dor1s) [reply infra/base-images/base-runner/bad_build_check]: The codes are intentionally ignored in `test_all`, as we run checks in parallel and make a judgement based on the percentage of broken fuzz targets https://github.com/google/oss-fuzz/blob/4d3073292ae43d99b1bd34b58e45021cb83760ae/infra/base-images/base-runner/test_all#L85 -- added a comment in `test_all`

When `bad_build_check` is used directly, e.g. for testing an individual fuzz target https://github.com/google/oss-fuzz/blob/4d3073292ae43d99b1bd34b58e45021cb83760ae/infra/helper.py#L505

The exit code is used to decide whether the check has failed or not https://github.com/google/oss-fuzz/blob/4d3073292ae43d99b1bd34b58e45021cb83760ae/infra/helper.py#L514

So I think it's all good, unless I'm missing anything else.

File: projects/zstd/project.yaml (hunk 4 above)
Comment (Dor1s) [reply projects/zstd/project.yaml]: That's a good point, I don't recall why I've added those here, must be a mistake. Moved coverage instrumentation flags into `COVERAGE_FLAGS_dataflow`, `-fsanitize=dataflow` is in `SANITIZER_FLAGS_dataflow`. Thanks!

File: infra/helper.py (hunk 11 above)
Comment (Dor1s) [reply infra/helper.py]: > If I lose this disagreement over whether a seperate engine exists then please fail loudly if --sanitizer dataflow is specified but engine isn't data flow.

Done, now re-building images and testing locally.

Comment (jonathanmetzman) [reply infra/helper.py]: > I think this won't be that useful, as I need to distinguish whether the argument is not used at all (the first condition) or is not specified (the second condition) and needs a default value.

Ah right my bad.

Comment (jonathanmetzman) [reply infra/helper.py]: > > If I lose this disagreement over whether a seperate engine exists then please fail loudly if --sanitizer dataflow is specified but engine isn't data flow.
> 
> Done, now re-building images and testing locally.

Thanks!

**Comment by Dor1s** [comment]
Alright, everything looks good to me. Thanks a lot @jonathanmetzman for the thorough review! I think I'm ready to merge, can you take another look please?

File: infra/base-images/base-builder/compile [hunk 12]
```diff
@@ -17,6 +17,11 @@
 
 echo "---------------------------------------------------------------"
 
+if [ "$SANITIZER" = "dataflow" ] && [ "$FUZZING_ENGINE" != "dataflow" ]; then
```
Comment (jonathanmetzman) [inline infra/base-images/base-builder/compile]: nit: Actually, I wonder if it is worth putting a check in helper.py so that user doesn't need to wait for image to be setup before failure happens.

Comment (Dor1s) [reply infra/base-images/base-builder/compile]: Was thinking about that and ended up deciding that to fail loudly it's better to fail for all possible scenarios, not only though the helper script.

**Comment by Dor1s** [comment]
Merging this and kicking off base images build, to make sure I monitor stuff and nothing's broken.
